src/main/java/org/folio/linked/data/integration/kafka/sender/search/AuthorityCreateMessageSender.java

```diff
@@ -35,13 +38,35 @@ public class AuthorityCreateMessageSender implements CreateMessageSender {
 
   @Override
   public Collection<Resource> apply(Resource resource) {
-    return Stream.of(resource)
-      .filter(this::test)
-      .toList();
+    return findAuthorities(resource, new HashSet<>());
   }
 
-  //TODO refactoring to extract resources
-  private boolean test(Resource resource) {
+  private Set<Resource> findAuthorities(Resource resource, Set<Resource> visitedResources) {
+    if (visitedResources.contains(resource)) {
+      return Collections.emptySet();
+    }
+    var authorities = new HashSet<Resource>();
+    if (isNewAuthority(resource)) {
+      authorities.add(resource);
+    }
+    visitedResources.add(resource);
+    resource.getOutgoingEdges().stream()
+      .map(ResourceEdge::getTarget)
+      .filter(target -> isNewAndNotVisited(target, visitedResources))
+      .flatMap(target -> findAuthorities(target, visitedResources).stream())
+      .forEach(authorities::add);
+    return authorities;
+  }
+
+  private boolean isNewAndNotVisited(Resource resource, Set<Resource> visited) {
+    return resource.isNew() && !visited.contains(resource);
```

Comment:
I was thinking `isNew()` will be false by the time this code is executed. Looks like I was totally wrong. I have to read more Spring :)